packages/obonode/obojobo-sections-assessment/server/models/assessment.js

```diff
@@ -0,0 +1,501 @@
+			{ userId, draftId, assessmentId, isPreview, resourceLinkId }
+		)
+		// @TODO: return hydrated Assessment objects
+	}
+
+	// Finish an attempt and write a new assessment score record
+	static completeAttempt(
+		assessmentId,
+		attemptId,
+		userId,
+		draftId,
+		draftContentId,
+		attemptScoreResult,
+		assessmentScoreDetails,
+		isPreview,
+		resourceLinkId
+	) {
+		return db
+			.tx(dbTransaction => {
+				const q1 = dbTransaction.one(
+					`
+					UPDATE attempts
+					SET
+						completed_at = now(),
+						result = $[result]
+					WHERE id = $[attemptId]
+					RETURNING
+						id AS "attemptId",
+						created_at as "startTime",
+						completed_at as "endTime",
+						assessment_id as "assessmentId",
+						state,
+						result as "scores"
+				`,
+					{ result: attemptScoreResult, attemptId: attemptId }
+				)
+
+				const aScore = new AssessmentScore({
+					userId,
+					draftId,
+					draftContentId,
+					assessmentId,
+					attemptId,
+					score: assessmentScoreDetails.assessmentModdedScore,
+					scoreDetails: assessmentScoreDetails,
+					isPreview,
+					resourceLinkId
+				})
+
+				const q2 = aScore.create(dbTransaction)
+
+				return dbTransaction.batch([q1, q2])
+			})
+			.then(result => ({
+				attemptData: result[0],
+				assessmentScoreId: result[1].id
+			}))
+	}
+
+	static fetchAttempts(userId, draftId, isPreview, resourceLinkId, optionalAssessmentId = null) {
+		return db
+			.manyOrNone(
+				`
+				SELECT
+					ROW_NUMBER () OVER (
+						PARTITION by ATT.assessment_id
+						ORDER BY ATT.completed_at, ATT.created_at
+					) AS "attempt_number",
+					ATT.*,
+					SCO.id AS "assessment_score_id",
+					SCO.score AS "assessment_score",
+					SCO.score_details AS "score_details"
+				FROM attempts ATT
+				LEFT JOIN assessment_scores SCO
+					ON ATT.id = SCO.attempt_id
+				WHERE
+					ATT.user_id = $[userId]
+					AND ATT.draft_id = $[draftId]
+					AND ATT.resource_link_id = $[resourceLinkId]
+					${optionalAssessmentId !== null ? 'AND ATT.assessment_id = $[optionalAssessmentId]' : ''}
+					AND ATT.is_preview = $[isPreview]
+				ORDER BY ATT.completed_at`,
+				{
+					userId,
+					draftId,
+					optionalAssessmentId,
+					isPreview,
+					resourceLinkId
+				}
+			)
+			.then(attempts => attempts.map(attempt => new AssessmentModel(attempt)))
+	}
+
+	static getAttemptIdsForUserForDraft(userId, draftId, resourceLinkId, isPreview) {
+		// Note, this must use the same sorting as getAttempts()
+		// for the attempt_number to be predictable
+		return db.manyOrNone(
+			`
+			SELECT
+			ROW_NUMBER () OVER (
+				PARTITION by assessment_id
+				ORDER BY completed_at, created_at
+			) AS "attempt_number",
+			id
+			FROM attempts
+			WHERE
+				user_id = $[userId]
+			AND draft_id = $[draftId]
+			AND resource_link_id = $[resourceLinkId]
+			AND is_preview = $[isPreview]
+			ORDER BY completed_at
+			`,
+			{ userId, draftId, isPreview, resourceLinkId }
+		)
+	}
+
+	static fetchAttemptByID(attemptId) {
```

Comment:
We're using `...Id` elsewhere

```suggestion
	static fetchAttemptById(attemptId) {
```

Reply:
There's gonna be a handful of these, fixed in commit 38c5de3c51be99f3219a418acd9d92a58e36c71a

Reply:
Updated